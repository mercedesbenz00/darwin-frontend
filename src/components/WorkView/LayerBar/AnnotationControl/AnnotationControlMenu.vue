<template>
  <div class="annotation-control-menu">
    <delete-all-visible-annotations
      v-if="!readonly && !isVersion2"
      class="annotation-control__delete"
      @close="$emit('close')"
    />
    <hide-by-class class="annotation-control__classes" />
    <hide-by-author class="annotation-control__authors" />
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, watch } from 'vue'

import { useStore } from '@/composables'
import { StageAnnotation } from '@/store/modules/workview/types'

import DeleteAllVisibleAnnotations from './DeleteAllVisibleAnnotations.vue'
import HideByAuthor from './HideByAuthor/HideByAuthor.vue'
import HideByClass from './HideByClass/HideByClass.vue'

export default defineComponent({
  name: 'AnnotationControlMenu',
  components: {
    DeleteAllVisibleAnnotations,
    HideByAuthor,
    HideByClass
  },
  props: {
    readonly: { required: false, default: false, type: Boolean }
  },
  setup () {
    const { getters, dispatch } = useStore()

    const isVersion2 = computed<boolean>(() => getters['dataset/isVersion2'])
    const nonTagAnnotations = computed<StageAnnotation[]>(
      () => getters['workview/selectedStageNonTagAnnotations']
    )

    watch(
      nonTagAnnotations,
      () => {
        dispatch('workview/refreshAnnotationsVisibility')
      }
    )

    return {
      isVersion2
    }
  }
})
</script>

<style lang="scss" scoped>
.annotation-control-menu {
  width: 225px;
  max-height: 80vh;
  overflow-y: auto;
  border-radius: 3px;
  padding: 7px 0;
}

.annotation-control__delete {
  margin-bottom: 5px;
}

.annotation-control__classes {
  margin-bottom: 5px;
}

.annotation-control__authors {
  margin-bottom: 5px;
}
</style>
