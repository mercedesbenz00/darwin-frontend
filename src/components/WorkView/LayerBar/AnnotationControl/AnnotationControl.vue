<template>
  <div class="annotation-control">
    <v-popover
      placement="left-start"
      popover-class="tooltip--annotation-control"
      trigger="manual"
      :open.sync="open"
    >
      <button
        class="annotation-control__trigger"
        @click="open = true"
      >
        <dots-vertical />
      </button>
      <template #popover>
        <annotation-control-menu
          :readonly="readonly"
          @close="open = false"
        />
      </template>
    </v-popover>

    <delete-confirmation-dialog
      v-if="!readonly"
      button-text="DELETE"
      name="confirm-delete-visible-annotations"
      title="Delete All Visible Annotations"
      detail="Are you sure? This will remove all visible annotations present on this image."
      @confirmed="deleteAllVisibleAnnotations"
    />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import DotsVertical from '@/assets/icons/V2/Mono/dots-vertical.vue'
import { DeleteConfirmationDialog } from '@/components/Common/DeleteConfirmationDialog/V1'

import AnnotationControlMenu from './AnnotationControlMenu.vue'

@Component({
  name: 'annotation-control',
  components: { AnnotationControlMenu, DeleteConfirmationDialog, DotsVertical }
})
export default class AnnotationControl extends Vue {
  @Prop({ required: false, default: false, type: Boolean })
  readonly!: boolean

  open: boolean = false

  async deleteAllVisibleAnnotations () {
    this.$modal.hide('confirm-delete-visible-annotations')
    await this.$store.dispatch('workview/deleteAllVisibleStageAnnotations')
  }
}
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.tooltip--annotation-control {
  @include dropdownBelow;
  border-radius: 3px;

  .tooltip-inner {
    background-color: $colorWhite;
    color: $colorSecondaryDark1;
    max-width: 250px;
    padding: 0;
    border-radius: 3px;
    overflow: hidden;
  }
}
</style>

<style lang="scss" scoped>
.annotation-control__trigger {
  @include row--center;
  background: transparent;
  position: relative;
  padding: 4px 5px;
  width: 25px;
  height: 25px;
  border-radius: 50%;
  transition: background .2s ease;

  &:hover {
    background: $colorAliceShade;
  }

  &:active {
    background: $colorAliceShadow;
  }
}
</style>
