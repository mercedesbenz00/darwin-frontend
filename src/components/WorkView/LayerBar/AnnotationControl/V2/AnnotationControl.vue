<template>
  <div class="annotation-control">
    <v-popover
      placement="left-start"
      popover-class="tooltip--annotation-control"
      trigger="manual"
      :open.sync="open"
    >
      <button
        class="annotation-control__trigger"
        @click="open = true"
      >
        <dots-vertical />
      </button>
      <template #popover>
        <annotation-control-menu
          :readonly="readonly"
          @close="open = false"
        />
      </template>
    </v-popover>

    <delete-confirmation-dialog
      v-if="!readonly"
      button-text="DELETE"
      name="confirm-delete-visible-annotations"
      title="Delete All Visible Annotations"
      detail="Are you sure? This will remove all visible annotations present on this image."
      @confirmed="deleteAllVisibleAnnotations"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue'

import DotsVertical from '@/assets/icons/V2/Mono/dots-vertical.vue'
import { DeleteConfirmationDialog } from '@/components/Common/DeleteConfirmationDialog/V1'
import { useModal } from '@/composables'
import { useActiveView } from '@/composables/useEditorV2'

import AnnotationControlMenu from './AnnotationControlMenu.vue'

export default defineComponent({
  name: 'AnnotationControl',
  components: { AnnotationControlMenu, DeleteConfirmationDialog, DotsVertical },
  props: {
    readonly: { required: false, default: false, type: Boolean }
  },
  setup () {
    const modal = useModal()
    const activeView = useActiveView()

    const open = ref(false)

    const deleteAllVisibleAnnotations = (): void => {
      modal.hide('confirm-delete-visible-annotations')
      activeView.value.annotationManager.annotations.forEach(a => {
        if (!a.isVisible) { return }
        activeView.value.annotationManager.deleteAnnotation(a)
      })
    }

    return {
      open,
      deleteAllVisibleAnnotations
    }
  }
})
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.tooltip--annotation-control {
  @include dropdownBelow;
  border-radius: 3px;

  .tooltip-inner {
    background-color: $colorWhite;
    color: $colorSecondaryDark1;
    max-width: 250px;
    padding: 0;
    border-radius: 3px;
    overflow: hidden;
  }
}
</style>

<style lang="scss" scoped>
.annotation-control__trigger {
  @include row--center;
  background: transparent;
  position: relative;
  padding: 4px 5px;
  width: 25px;
  height: 25px;
  border-radius: 50%;
  transition: background .2s ease;

  &:hover {
    background: $colorAliceShade;
  }

  &:active {
    background: $colorAliceShadow;
  }
}
</style>
