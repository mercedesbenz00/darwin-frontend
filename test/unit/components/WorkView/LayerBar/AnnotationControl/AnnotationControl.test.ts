import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { stubDirectiveWithAttribute } from 'test/unit/directiveStubs'
import { VPopover } from 'test/unit/stubs'

import AnnotationControl from '@/components/WorkView/LayerBar/AnnotationControl/AnnotationControl.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('tooltip', stubDirectiveWithAttribute)

let store: ReturnType<typeof createTestStore>

const mocks = { $modal: { hide: jest.fn() } }
const stubs = { VPopover }
let propsData: { readonly?: boolean }

beforeEach(() => {
  store = createTestStore()
  propsData = {}
})

it('matches snapshot', () => {
  const wrapper = shallowMount(AnnotationControl, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when readonly', () => {
  propsData.readonly = true
  const wrapper = shallowMount(AnnotationControl, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('dispatches workview/deleteAllVisibleStageAnnotations when delete is confirmed', async () => {
  const wrapper = shallowMount(AnnotationControl, { localVue, mocks, propsData, store, stubs })
  await wrapper.find('delete-confirmation-dialog-stub').vm.$emit('confirmed')
  expect(store.dispatch).toBeCalledWith('workview/deleteAllVisibleStageAnnotations')
})
