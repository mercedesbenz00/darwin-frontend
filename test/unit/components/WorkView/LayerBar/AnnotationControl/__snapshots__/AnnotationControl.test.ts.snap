// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="annotation-control">
  <div class="v-popover-stub" placement="left-start" popover-class="tooltip--annotation-control" trigger="manual">
    <div class="main"><button class="annotation-control__trigger">
        <dots-vertical-stub></dots-vertical-stub>
      </button></div>
    <div class="popover">
      <annotation-control-menu-stub></annotation-control-menu-stub>
    </div>
  </div>
  <delete-confirmation-dialog-stub title="Delete All Visible Annotations" detail="Are you sure? This will remove all visible annotations present on this image." buttontext="DELETE" name="confirm-delete-visible-annotations"></delete-confirmation-dialog-stub>
</div>
`;

exports[`matches snapshot when readonly 1`] = `
<div class="annotation-control">
  <div class="v-popover-stub" placement="left-start" popover-class="tooltip--annotation-control" trigger="manual">
    <div class="main"><button class="annotation-control__trigger">
        <dots-vertical-stub></dots-vertical-stub>
      </button></div>
    <div class="popover">
      <annotation-control-menu-stub readonly="true"></annotation-control-menu-stub>
    </div>
  </div>
  <!---->
</div>
`;
